src/inmanta/execute/dataflow.py

```diff
@@ -285,72 +596,358 @@ class AssignableNode(Node):
+        self.equivalence.propagate_tentative_instance()
+
+    def assign_assignable(self, var_ref: AssignableNodeReference, responsible: "Locatable", context: "DataflowGraph") -> None:
+        """
+            Assigns an assignable node to this node, by reference.
+        """
+        # Gather all equivalences on the path between the rhs's leaves and this node.
+        # The existence of such a trail indicates an assignment loop.
+        equivalence_trail: Set[Equivalence] = reduce(
+            set.union, (n.equivalence.equivalences_on_path(self) for n in var_ref.nodes()), set(),
+        )
+        # merge all equivalences on the trail
+        new_equivalence: Equivalence = reduce(
+            lambda acc, eq: acc.merge(eq), equivalence_trail, Equivalence(frozenset()),
+        )
+        self.assignable_assignments.append(Assignment(self, var_ref, responsible, context))
+        # apply new equivalence to all it's nodes
+        for node in new_equivalence.nodes:
+            node.equivalence = new_equivalence
+        # propagate this node's tentative instance to the new leaves, if it exists
+        self.equivalence.propagate_tentative_instance()
+
+
+class Equivalence:
+    """
+        Represents a collection of nodes that are equivalent because of one or more assignment loops.
+    """
+
+    __slots__ = ("nodes", "tentative_instance")
+
+    def __init__(
+        self, nodes: FrozenSet[AssignableNode] = frozenset(), tentative_instance: Optional["InstanceNode"] = None
+    ) -> None:
+        self.nodes: FrozenSet[AssignableNode] = nodes
+        self.tentative_instance: Optional[InstanceNode] = tentative_instance
+
+    def merge(self, other: "Equivalence") -> "Equivalence":
+        """
+            Returns the equivalence that is the union of this one and the one passed as an argument.
+        """
+        tentative_instance: Optional[InstanceNode] = self.tentative_instance
+        if tentative_instance is not None:
+            self.tentative_instance = None
+            if other.tentative_instance is not None:
+                tentative_instance.merge(other.tentative_instance)
+        return Equivalence(self.nodes.union(other.nodes), tentative_instance)
+
+    def is_leaf(self) -> bool:
+        """
+            Returns true iff this equivalence is a leaf. An equivalence is a leaf iff it has an instance
+            or value assignment, or doesn't have any assignments at all.
+        """
+        return any(self.instance_assignments()) or any(self.value_assignments()) or not any(self.assignable_assignments())
+
+    def leaves(self) -> Iterator[AssignableNodeReference]:
+        """
+            Returns an iterator over all leaves on assignment paths originating from this equivalence.
+        """
+        if self.is_leaf():
+            explicit_leaves: Iterator[AssignableNodeReference] = (
+                node.reference() for node in self.nodes if len(node.value_assignments) > 0 or len(node.instance_assignments) > 0
```

Comment:
wouldn't this and is_leaf be cleaner if the nodes have is_leaf as well?

Reply:
The problem with that approach is that it's difficult to tell if a node is a leaf without looking at the equivalence, since outgoing edges could just be part of a loop. At equivalence level we can filter only the external assignments. We could provide an is_leaf method but it would have to do the same filtering that is done in `Equivalence.assignable_assignments()`. I believe such filtering belongs in the Equivalence, not the Node. Do you agree?

Reply:
I would expect is_leaf to return len(node.value_assignments) > 0 or len(node.instance_assignments) > 0

Reply:
(otherwise I would document it)

Reply:
That would indeed delegate something to the Node, but then the semantics of what a leaf is would differ between AssignableNode and Equivalence since the Node method does not take edges to other AssignableNodes into account. Maybe a compromise where we put the method in AssignableNode but rename it to something that doesn't contain the word leaf since it wouldn't detect all leaves?

Reply:
or leave it as it is.